### PR TITLE
prevent not-allowed cpu recognize

### PR DESCRIPTION
**Pull request by Sunmon**
issue #7157
Signed-off-by : hansolpp <hansolmin3@gmail.com>
Signed-off-by: sunmon <blackmilk274@gmail.com>

This is the problem:

```bash
docker run --cpuset-cpus="1,3" [openmpi image]

# on mpi
mpirun -np 4 [example.out] 
```

mpirun must not be runned, because container has 2 cpus but np gives 4 processes. So it has to be print 'slot error', but it doesn't.

So I fixed this problem by prevent mpi recognizing unallocated cpu.
But it is only temporary solution.

### CODE REVIEW

**Comment by awlauria** [comment]
ok to test

File: opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c [hunk 1]
```diff
@@ -3419,6 +3422,86 @@ look_sysfsnode(struct hwloc_topology *topology,
...
+
+        last_explore = current_explore;
+        set_allowed(allowed, num1, num2);
+    }
+}
+
+void read_cpuset_cpus(char* buf, char* path)
+{
+    FILE *fd = fopen(path, "r");
+    fgets(buf, 20, fd);
```
Comment (awlauria) [inline opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c]: if you get all 20, you won't have room for a `'0' in `buf`

**Comment by rhc54** [comment]
Time out, folks. We are not going to modify the embedded hwloc in OMPI. We should discuss this with @bgoglin to see if (a) there already is a solution in hwloc (I believe there is, but we may not be using it) and (b) what the correct solution should be.

File: opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c [hunk 2]
```diff
@@ -3419,6 +3422,86 @@ look_sysfsnode(struct hwloc_topology *topology,
...
+{
+    char *last_explore, *current_explore;
+
+    current_explore = buf;
+    last_explore = current_explore; //last_explore는 마지막으로 탐색한 숫자 첫 자리에 위치
+
+    while( *(++current_explore) != '\n' && *current_explore != '\0')
+    {
+        current_explore = find_next_deli(current_explore);
+        if( current_explore == NULL) current_explore = buf + strlen(buf) - 1;
```
Comment (awlauria) [inline opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c]: I'm not sure what the purpose of this line is.

if current_explore is NULL, why not break out of the loop?

File: opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c [hunk 3]
```diff
@@ -3443,6 +3526,10 @@ look_sysfscpu(struct hwloc_topology *topology,
     struct dirent *dirent;
     cpuset = hwloc_bitmap_alloc();
 
+    int allow[20];
```
Comment (awlauria) [inline opal/mca/hwloc/hwloc201/hwloc/hwloc/topology-linux.c]: does this only allow for 20 cpus per node?

**Comment by rhc54** [comment]
I've flagged Brice on the issue that was filed. Let's table this PR for the moment until we see how he wants to proceed.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3ff99f228b7e99a72c6f6282038c3368

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5d8bd98770e950a214fd47f30a69085b

**Comment by bgoglin** [comment]
@Sunmon hwloc already supports cgroups, so things should work fine. One thing that isn't supported in current releases is Linux cgroups v2 (I added it to git a couple weeks ago), but your problem seems to involve cgroups v1 only.
Can you run hwloc-gather-topology inside that docker and send the generated tarball to me (brice.goglin@inria.fr) so that I try to debug why it fails here?
Otherwise the output of lstopo when hwloc is configured with --enable-debug would also be helpful.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e17d844b6e88682144e7e43a4969b7ff

**Pull request closed without merging**